### PR TITLE
Add `ErrorGuaranteed` to `hir::{Expr,Ty}Kind::Err` variants

### PR DESCRIPTION
**Pull request by compiler-errors**
First step in making the `Err` variants of `ExprKind` and `TyKind` require an `ErrorGuaranteed` during parsing. Making the corresponding AST versions require `ErrorGuaranteed` is a bit harder, whereas it was pretty easy to do this for HIR, so let's do that first.

The only weird thing about this PR is that `ErrorGuaranteed` is moved to `rustc_span`. This is *certainly* not the right place to put it, but `rustc_hir` cannot depend on `rustc_error` because the latter already depends on the former. Should I just pull out some of the error machinery from `rustc_error` into an even more minimal crate that `rustc_hir` can depend on? Advice would be appreciated.

### CODE REVIEW

File: compiler/rustc_span/src/lib.rs [hunk 1]
```diff
@@ -2149,3 +2149,17 @@ where
         Hash::hash(&len, hasher);
     }
 }
+
+/// Useful type to use with `Result<>` indicate that an error has already
```
Comment (compiler-errors) [inline compiler/rustc_span/src/lib.rs:2153]: As I noted in the PR description, this shouldn't live here -- it's just here as a proof of concept for the rest of the PR.

Where should this live instead? :thinking:

**Comment by cjgillot** [comment]
> The only weird thing about this PR is that `ErrorGuaranteed` is moved to `rustc_span`. This is _certainly_ not the right place to put it, but `rustc_hir` cannot depend on `rustc_error` because the latter already depends on the former. Should I just pull out some of the error machinery from `rustc_error` into an even more minimal crate that `rustc_hir` can depend on? Advice would be appreciated.

Why does `rustc_errors` depend on `rustc_hir`? Can the dependency chain become `rust_span -> rustc_errors -> rustc_hir` instead by moving stuff out of `rustc_hir`?

**Comment by compiler-errors** [comment]
@cjgillot:

> Why does `rustc_errors` depend on `rustc_hir`?

Two reasons. First is that `rustc_errors` depends on `rustc_hir` directly, but that can be easily removed. Second is because `rustc_errors` depends on `rustc_lint_defs` (because we use `LintExpectationId` in diagnostics), and `rustc_lint_defs` depends on `rustc_hir` because `LintExpectationId` has a `HirId` inside of it.

Even if I fix this somehow, I'll be back into the same situation if/when I want to add `ErrorGuaranteed` to `ast::ExprKind::Err`, etc., because `LintExpectationId` also depends on `ast::AttrId`.

I'm not sure if there's a good way to untangle this cycle now I think of it. I could pull things out into a separate crate?

**Comment by cjgillot** [comment]
What a knot, indeed. `ErrorGuaranteed` could move to `rustc_span`, and be accessible to all from there.
`rustc_errors` already uses `unchecked_claim_error_was_emitted`, so that wouldn't break much of it.

**Comment by cjgillot** [comment]
r=me after rebase

**Pull request merged**